manual/src/cmds/runtime-tracing.etex

```diff
@@ -0,0 +1,266 @@
+
+The full set of events emitted by probes and their documentation can be found in
+\ifouthtml
+ \ahref{libref/Runtime_events.html}{Module \texttt{Runtime_events}}.
+\else
+ section~\ref{Runtime_events}.
+\fi
+
+\subsection{s:runtime-tracing-ingestion}{Events transport}
+
+The second part of the system ingests events emitted by the probes and makes
+these available to consumers.
+
+\subsubsection{s:runtime-tracing-ringbuffers}{Ring buffers}
+
+Events are transported using a data structure known as a {\em ring buffer}. This
+data structure consists of two pointers into a linear backing array, the tail
+pointer points to a location where new events can be written and the head
+pointer points to the oldest event in the buffer that can be read. When
+insufficient space is available in the backing array to write new events, the
+head pointer is advanced and the oldest events are overwritten by new ones.
+
+The ring buffer implementation used in runtime events can be written by at most
+one producer at a time but can be read simultaneously by multiple consumers
+without coordination from the producer. There is a unique ring buffer for every
+running domain and on domain termination ring buffers may be re-used for newly
+spawned domains. The ring buffers themselves are stored in a memory-mapped file
+with the processes identifier as the name and the extension ".events", this
+enables them to be read from outside the main OCaml process.
+
+\subsubsection{s:runtime-tracing-apis}{Consumption APIs}
+
+Runtime events consists of both OCaml and C APIs which are cursor-based and
+polling-driven. The high-level process for consuming events is as follows:
+
+\begin{enumerate}
+  \item A cursor is created via Runtime_events.create_cursor for either the
+  current process or an external process (specified by a path and PID).
+  \item A callbacks record is created via Runtime_events.Callbacks.create with
+  the desired callback functions.
+  \item The cursor is polled via Runtime_events.read_poll using the callbacks
+  created in the previous step. Any supplied callback functions are called for
+  each matching event in the ring buffers.
+\end{enumerate}
+
+\section{s-runtime-tracing-usage}{Usage}
+
+\subsection{s-runtime-tracing-ocaml-apis}{With OCaml APIs}
+
+We start with a simple example that prints the name, begin and end times
+for events:
+
+\begin{verbatim}
+let runtime_begin _ ts phase =
+    Printf.printf "Begin\t%s\t%Ld\n"
+        (Runtime_events.runtime_phase_name phase)
+        (Runtime_events.Timestamp.to_int64 ts)
+
+let runtime_end _ ts phase =
+    Printf.printf "End\t%s\t%Ld\n"
+        (Runtime_events.runtime_phase_name phase)
+        (Runtime_events.Timestamp.to_int64 ts)
+
+let () =
+    Runtime_events.start ();
+    let cursor = Runtime_events.create_cursor None in
+    let callbacks = Runtime_events.Callbacks.create ~runtime_begin ~runtime_end ()
+    in
+    while true do
+        let list_ref = ref [] in (* for later fake work *)
```

Comment:
"for later fake work" ?

Reply:
clarified this is fake GC work: https://github.com/ocaml/ocaml/pull/11349/commits/a1746bfeca44c13422d23efb078e43def8da8bdb